test/src/ui/widget.spec.ts

```diff
@@ -0,0 +1,1792 @@
+        child.parent = parent1;
+        expect(parent0.messages.indexOf('child-removed')).to.not.be(-1);
+        expect(parent1.messages.indexOf('child-added')).to.not.be(-1);
+      });
+
+      it('should throw an error if the widget contains the parent', () => {
+        let widget0 = new Widget();
+        let widget1 = new Widget();
+        widget0.parent = widget1;
+        expect(() => { widget1.parent = widget0; }).to.throwError();
+      });
+
+      it('should be a no-op if there is no parent change', () => {
+        let parent = new LogWidget();
+        let child = new Widget();
+        child.parent = parent;
+        child.parent = parent;
+        expect(parent.messages.indexOf('child-removed')).to.be(-1);
+      });
+
+    });
+
+    describe('#layout', () => {
+
+      it('should default to `null`', () => {
+        let widget = new Widget();
+        expect(widget.layout).to.be(null);
+      });
+
+      it('should set the layout for the widget', () => {
+        let widget = new Widget();
+        let layout = new LogLayout();
+        widget.layout = layout;
+        expect(widget.layout).to.be(layout);
+      });
+
+      it('should be single-use only', () => {
+        let widget = new Widget();
+        widget.layout = new LogLayout();
+        expect(() => { widget.layout = new LogLayout(); }).to.throwError();
+      });
+
+      it('should be disposed when the widget is disposed', () => {
+        let widget = new Widget();
+        let layout = new LogLayout();
+        widget.layout = layout;
+        widget.dispose();
+        expect(layout.isDisposed).to.be(true);
+      });
+
+      it('should be a no-op if the layout is the same', () => {
+        let widget = new Widget();
+        let layout = new LogLayout();
+        widget.layout = layout;
+        widget.layout = layout;
+        expect(widget.layout).to.be(layout);
+      });
+
+      it('should throw an error if the layout already has a parent', () => {
+        let widget0 = new Widget();
+        let widget1 = new Widget();
+        let layout = new LogLayout();
+        widget0.layout = layout;
+        expect(() => { widget1.layout = layout; }).to.throwError();
+      });
+
+      it('should throw an error if the `DisallowLayout` flag is set', () => {
+        let widget = new Widget();
+        widget.setFlag(WidgetFlag.DisallowLayout);
+        let layout = new LogLayout();
+        expect(() => { widget.layout = layout; }).to.throwError();
+      });
+
+    });
+
+    describe('#children()', () => {
+
+      it("should return an iterator over the widget's children", () => {
+        let widget = new Widget();
+        widget.layout = new LogLayout();
+        let child = widget.children().next();
+        expect(child instanceof LogWidget).to.be(true);
+      });
+
+      it('should return an empty iterator if there is no layout', () => {
+        let widget = new Widget();
+        expect(widget.children()).to.be(EmptyIterator.instance);
```

Comment:
this is testing an implementation detail. Better to exhaust the iterator into an array, and test that the array length is zero:

`expect(toArray(widget.children()).to.eql([])`